File: lib/platform/notifications.ts
/**
 * Desktop notification helper.
 *
 * On desktop (Tauri): sends native OS notifications via the notification plugin.
 * On web: falls back to the browser Notification API (if permitted).
 *
 * Usage:
 *   import { notify } from "@/lib/platform/notifications";
 *   await notify({ title: "New message", body: "From @alice", channel: "telegram" });
 */

"use client";

import { isDesktop } from "./index";
import { invoke } from "./tauri-invoke";

interface NotifyOptions {
  title: string;
  body: string;
}

/** Send a notification using the best available method for the current platform. */
export async function notify(options: NotifyOptions): Promise<void> {
  if (isDesktop) {
    await invoke("send_notification", {
      payload: {
        title: options.title,
        body: options.body,
      },
    });
    return;
  }

  // Web fallback: browser Notification API
  if (typeof Notification === "undefined") return;
  if (Notification.permission === "granted") {
    new Notification(options.title, { body: options.body });
  }
}

/** Check if notification permission is granted. */
export async function hasNotificationPermission(): Promise<boolean> {
  if (isDesktop) {
    return invoke<boolean>("check_notification_permission");
  }

  if (typeof Notification === "undefined") return false;
  return Notification.permission === "granted";
}

/** Request notification permission from the user. */
export async function requestNotificationPermission(): Promise<boolean> {
  if (isDesktop) {
    return invoke<boolean>("request_notification_permission");
  }

  if (typeof Notification === "undefined") return false;
  const result = await Notification.requestPermission();
  return result === "granted";
}

File: desktop/src-tauri/src/notifications.rs
//! Native OS notifications for desktop.
//!
//! Sends system notifications for Telegram messages, emails, deal updates, etc.
//! Uses Tauri's notification plugin for cross-platform support.

use serde::Deserialize;
use tauri::command;
use tauri_plugin_notification::NotificationExt;

#[derive(Deserialize)]
pub struct NotificationPayload {
    pub title: String,
    pub body: String,
}

/// Send a native OS notification.
#[command]
pub fn send_notification(
    app: tauri::AppHandle,
    payload: NotificationPayload,
) -> Result<(), String> {
    app.notification()
        .builder()
        .title(&payload.title)
        .body(&payload.body)
        .show()
        .map_err(|e| e.to_string())
}

/// Check if notification permission is granted.
#[command]
pub fn check_notification_permission(app: tauri::AppHandle) -> Result<bool, String> {
    let perm = app
        .notification()
        .permission_state()
        .map_err(|e| e.to_string())?;
    Ok(perm == tauri_plugin_notification::PermissionState::Granted)
}

/// Request notification permission from the OS.
#[command]
pub fn request_notification_permission(app: tauri::AppHandle) -> Result<bool, String> {
    let perm = app
        .notification()
        .request_permission()
        .map_err(|e| e.to_string())?;
    Ok(perm == tauri_plugin_notification::PermissionState::Granted)
}
